File: tests/invariants/lib/inputs-hash.ts
import { createHash } from 'node:crypto';

/**
 * Mirror of planner/inputs_hash.go HashInputs (I-02 reuse key).
 * WHY: the reuse-sensitivity property ("any output-affecting input change
 * ⇒ new key") is property-tested against this independent implementation.
 */

export interface InputsMaterial {
  baseSha: string;
  lockfiles: readonly string[];
  flags: readonly string[];
  toolchain: string;
}

function sorted(values: readonly string[]): string[] {
  return [...values].sort();
}

export function hashInputs(m: InputsMaterial): string {
  const material = [
    `base_sha=${m.baseSha}`,
    `lockfiles=${sorted(m.lockfiles).join(',')}`,
    `flags=${sorted(m.flags).join(',')}`,
    `toolchain=${m.toolchain}`,
  ].join('\n');
  return 'sha256:' + createHash('sha256').update(material, 'utf8').digest('hex');
}

/** The documented I-02 rule: reuse is legal only on FULL equality of every field. */
export function materialsEqualForReuse(a: InputsMaterial, b: InputsMaterial): boolean {
  const sameSet = (x: readonly string[], y: readonly string[]) =>
    x.length === y.length && sorted(x).every((v, i) => v === sorted(y)[i]);
  return (
    a.baseSha === b.baseSha &&
    a.toolchain === b.toolchain &&
    sameSet(a.lockfiles, b.lockfiles) &&
    sameSet(a.flags, b.flags)
  );
}

File: services/control-plane/internal/planner/inputs_hash.go
package planner

import (
	"crypto/sha256"
	"encoding/hex"
	"sort"
	"strings"
)

// InputsMaterial is the full reuse key per I-02: base SHA, lockfiles, flags
// and toolchain. Every field participates in the digest.
type InputsMaterial struct {
	BaseSHA   string
	Lockfiles []string
	Flags     []string
	Toolchain string
}

// HashInputs computes the plan-level inputs_hash: "sha256:<hex>" over a
// canonical serialization of the material. Slices are sorted so equivalent
// inputs hash identically regardless of submission order; a change in any
// component changes the digest (changed input ⇒ miss).
func HashInputs(m InputsMaterial) string {
	lockfiles := append([]string(nil), m.Lockfiles...)
	flags := append([]string(nil), m.Flags...)
	sort.Strings(lockfiles)
	sort.Strings(flags)
	var b strings.Builder
	b.WriteString("base_sha=")
	b.WriteString(m.BaseSHA)
	b.WriteString("\nlockfiles=")
	b.WriteString(strings.Join(lockfiles, ","))
	b.WriteString("\nflags=")
	b.WriteString(strings.Join(flags, ","))
	b.WriteString("\ntoolchain=")
	b.WriteString(m.Toolchain)
	sum := sha256.Sum256([]byte(b.String()))
	return "sha256:" + hex.EncodeToString(sum[:])
}
